src/mbgl/util/i18n.cpp

```diff
@@ -352,6 +374,188 @@ bool allowsIdeographicBreaking(uint16_t chr) {
     //        || isInCJKCompatibilityIdeographsSupplement(chr));
 }
 
+bool allowsVerticalWritingMode(const std::u16string& string) {
+    for (char32_t chr : string) {
+        if (hasUprightVerticalOrientation(chr)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+// The following logic comes from
+// <http://www.unicode.org/Public/vertical/revision-16/VerticalOrientation-16.txt>.
+// The data file denotes with “U” or “Tu” any codepoint that may be drawn
+// upright in vertical text but does not distinguish between upright and
+// “neutral” characters.
+
+bool hasUprightVerticalOrientation(char16_t chr) {
+    if (chr == u'˪' || chr == u'˫')
+        return true;
+
+    // Return early for characters outside all ranges whose characters remain
+    // upright in vertical writing mode.
+    if (chr < 0x1100)
+        return false;
+
+    if (isInBopomofo(chr) || isInBopomofoExtended(chr))
+        return true;
+    if (isInCJKCompatibilityForms(chr)) {
+        if (!(chr >= u'﹉' && chr <= u'﹏'))
+            return true;
+    }
+    if (isInCJKCompatibility(chr) || isInCJKCompatibilityIdeographs(chr) ||
+        isInCJKRadicalsSupplement(chr) || isInCJKStrokes(chr))
+        return true;
+    if (isInCJKSymbolsandPunctuation(chr)) {
+        if (!(chr >= u'〈' && chr <= u'】') && !(chr >= u'〔' && chr <= u'〟') && chr != u'〰')
+            return true;
+    }
+    if (isInCJKUnifiedIdeographs(chr) || isInCJKUnifiedIdeographsExtensionA(chr) ||
+        isInEnclosedCJKLettersandMonths(chr) || isInHangulCompatibilityJamo(chr) ||
+        isInHangulJamo(chr) || isInHangulJamoExtendedA(chr) || isInHangulJamoExtendedB(chr) ||
+        isInHangulSyllables(chr) || isInHiragana(chr) ||
+        isInIdeographicDescriptionCharacters(chr) || isInKanbun(chr) || isInKangxiRadicals(chr))
+        return true;
+    if (isInKatakana(chr)) {
+        if (chr != u'ー')
+            return true;
+    }
+    if (isInKatakanaPhoneticExtensions(chr))
+        return true;
+    if (isInHalfwidthandFullwidthForms(chr)) {
+        if (chr != u'（' && chr != u'）' && chr != u'－' && !(chr >= u'：' && chr <= u'＞') &&
+            chr != u'［' && chr != u'］' && chr != u'＿' && !(chr >= u'｛' && chr <= 0xFFDF) &&
+            chr != u'￣' && !(chr >= u'￨' && chr <= 0xFFEF))
+            return true;
+    }
+    if (isInSmallFormVariants(chr)) {
+        if (!(chr >= u'﹘' && chr <= u'﹞') && !(chr >= u'﹣' && chr <= u'﹦'))
+            return true;
+    }
+    if (isInUnifiedCanadianAboriginalSyllabics(chr) ||
+        isInUnifiedCanadianAboriginalSyllabicsExtended(chr) || isInVerticalForms(chr) ||
+        isInYijingHexagramSymbols(chr) || isInYiSyllables(chr) || isInYiRadicals(chr))
+        return true;
+
+    // https://github.com/mapbox/mapbox-gl/issues/29
+
+    // if (isInMeroiticHieroglyphs(chr)) return true;
```

Comment:
https://github.com/mapbox/mapbox-gl-js/pull/3608#discussion_r87851149 requested that I put the list of vertical supplementary blocks in mapbox/mapbox-gl#29, now mapbox/mapbox-gl-js#4001, instead of leaving them here as commented out lines.